### PR TITLE
refactor(datasource/npm): Convert to class

### LINKED ISSUE #8647: Refactor: Datasource classes
I would like to consider class-based datasources to achieve better DRY than today plus some enhance features.

Currently datasources are somewhat declaration-based and the datasource index then handles things, e.g. `caching`, default registryUrls, etc. This type of stuff should remain although maybe much of the datasource index's logic will move into the base datasource class.

We definitely want error handling shared as much as possible.

Caching would be another aspect done automatically.

Something new: our http layer returns an `auth` variable today to indicate that the request was auth'd. In future I think we may return an auth *fingerprint* (e.g. sha256 the authorization header content and then use the first 5 chars). This would then be used in conjunction with caching, to allow private lookup caching.

Need to discuss how to do this. Start small and then enhance it?

### PR DESCRIPTION
**Pull request by zharinov**
## Context:

- Closes: #8647

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 1]
```diff
@@ -63,7 +63,7 @@ describe('workers/repository/process/lookup/index', () => {
     it('returns rollback for pinned version', async () => {
       config.currentValue = '0.9.99';
       config.depName = 'q';
-      config.datasource = datasourceNpmId;
+      config.datasource = NpmDatasource.id;
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.spec.ts:66]: We should use a default config, so we don't need to repeat / change this line in so many places. Can be done in a follow up PR.

**Review by viceice — DISMISSED** [review]

**Comment by JamieMagee** [comment]
Couple of things:
1. Can you remove the `npm/common.ts` file now?
2. Caching in NPM needs a more thorough look (it's why I've been avoiding this datasource 😅)
3. Do you think it makes sense to combine into a single file class, like you did with docker? It can be done in another PR whatever the outcome.

**Comment by zharinov** [comment]
I was afraid changes this datasource too, so left this as close to the original as possible. That's why I didn't pursue any of changes you've listed.

**Comment by viceice** [comment]
Yeah, let's do other refactoring later.

**Review by viceice — APPROVED** [review]

**Pull request merged**